### ISSUE TITLE
How do I interpret evaluation results on my custom dataset?

### ISSUE THREAD

**Issue by Suvi-dha** [matterport/Mask_RCNN#663]
After performing training, I ran evaluation code on 500 images to which I received following results.
Can any one help me understand these results. Any help is much appreciated.  

```
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.061
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.075
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.059
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.094
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.099
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.099
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.099
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.149
```

**Comment by qmzsuxing** [comment]
@Suvi-dha have you solved your problem? I also wonder what does this mean...

**Comment by Suvi-dha** [comment]
yes, I read about it. The values as stated are average precision and average recall on IoU(intersection over union)  thresholds between 0.5 and 0.95  where maximum detections over which it is calculated is 100. In other words, precision is calculated over number of positive detections (greater than IoU threshold) by total number of TPs and FPs. and then the average of precision values over each threshold is calculated to obtain the final value in the last column. 
Similarly for recall. 
Also, they have calculated Average Precision (AP) and AR over different areas (all, small, medium and large)

**Comment by ash1995** [comment]
@Suvi-dha So how exactly would this result be reported?

**Comment by Suvi-dha** [comment]
MaskRCNN doesn't work on my dataset due to many issues, that's why the results are low. I tried training by changing hyperparameters to see the effect. So, you can also try changing them and then observe the results. May be they will improve. 
I don't completely understand  what do you mean by reporting the score of your model. 
If you are comparing your model performance with maskRCNN then the metrics you are using for reporting your result should also be calculated from MRCNN, or the other way round.

**Comment by ash1995** [comment]
@Suvi-dha By reporting the score I mean reporting the mAP score since that's the common evaluation metric. I tried calculating the mAP score using the compute_ap function as in the train_shapes.ipynb and that gave me a somewhat reasonable score of 0.272 or 27.2. I guess I'll just go with that.

**Comment by eyildiz-ugoe** [comment]
I am having exactly the same problem. What functions do you guys use to evaluate your model? Where are these functions located?

**Comment by banafsh89** [comment]
@Suvi-dha Could you please tell me which part of the code gives you Average Recall. I can only calculate AP using utils.compute_ap. I don't know how to calculate Average Recall!

**Comment by Suvi-dha** [comment]
> @Suvi-dha Could you please tell me which part of the code gives you Average Recall. I can only calculate AP using utils.compute_ap. I don't know how to calculate Average Recall!

you can calculate average recall from utils.compute_recall Also, compute_ap also returns iou, precision and recalls.

**Comment by Suvi-dha** [comment]
> I am having exactly the same problem. What functions do you guys use to evaluate your model? Where are these functions located? `train_shapes.ipynb` only has mAP, that's not enough. We need to evaluate it using the metrics given by COCO. (AP50, AP75, AP_S...etc)

Search in utils.py.

**Comment by banafsh89** [comment]
Thanks, but that is only recall in each iou range! how should I find the average recall! For example if you look at compute_ap, it finds mAP as "Compute mean AP over recall range". Do you have any idea bout that?

**Comment by Suvi-dha** [comment]
> Thanks, but that is only recall in each iou range! how should I find the average recall! For example if you look at compute_ap, it finds mAP as "Compute mean AP over recall range". Do you have any idea bout that?

you have to use pycocotools library if you want to calculate the results like as shown above in my first comment, on your dataset. 

> from pycocotools.coco import COCO
> from pycocotools.cocoeval import COCOeval
> from pycocotools import mask as maskUtils

check the code for 

> cocoeval

.

**Comment by sineagles** [comment]
Do you solve this problem? I meet the same problem same to yours. Can you tell me  how correct it?

**Comment by sineagles** [comment]
> Do you solve this problem? I meet the same problem same to yours. Can you tell me how correct it?

iii

**Comment by jdmelendez** [comment]
> @ Suvi-dha Entonces, ¿cómo se informará exactamente este resultado? ¿Es su mAP = 0.061? ¿No es este un valor extremadamente bajo? Me enfrento a un problema similar en el que mi puntaje AP para todos los umbrales de IOU es **0.075** . No entiendo cómo informar realmente el puntaje de mi modelo.

If you're training with COCO, the mAP score is the AP. Perhaps is this what you are looking for. Reading on [http://cocodataset.org/#detection-eval](url) , in "2.Metrics" there is the next line:

"AP is averaged over all categories. Traditionally, this is called "mean average precision" (mAP). We make no distinction between AP and mAP (and likewise AR and mAR) and assume the difference is clear from context."

![Captura1](https://user-images.githubusercontent.com/59265982/81265049-b1982180-9042-11ea-8cf7-43044eed273f.JPG)

**Comment by WillianaLeite** [comment]
> > @ Suvi-dha Você poderia me dizer qual parte do código fornece a média de recall? Só posso calcular o AP usando utils.compute_ap. Não sei como calcular o recall médio!
> 
> você pode calcular o recall médio a partir de utils.compute_recall Além disso, o compute_ap também retorna iou, precision e recalls.



> > @Suvi-dha Could you please tell me which part of the code gives you Average Recall. I can only calculate AP using utils.compute_ap. I don't know how to calculate Average Recall!
> 
> you can calculate average recall from utils.compute_recall Also, compute_ap also returns iou, precision and recalls.

Could you please explain to me how to calculate mAR from the "utils.compute_recall" function, I understand that it returns the AR, but how should I calculate the mAR? Please help me!!

**Comment by Joanne513** [comment]
> After performing training, I ran evaluation code on 500 images to which I received following results.
> Can any one help me understand these results. Any help is much appreciated.
> 
> ```
> Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.061
>  Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.075
>  Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.059
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.094
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.099
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.099
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.099
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.149
> ```

how you got this result? i only can use the compute_ap function to calcute the AP @ IoU = 0.5, how you got so many IoUs? can you help me!

**Comment by elia86** [comment]
> > > @ Suvi-dha Você poderia me dizer qual parte do código fornece a média de recall? Só posso calcular o AP usando utils.compute_ap. Não sei como calcular o recall médio!
> > 
> > 
> > você pode calcular o recall médio a partir de utils.compute_recall Além disso, o compute_ap também retorna iou, precision e recalls.
> 
> > > @Suvi-dha Could you please tell me which part of the code gives you Average Recall. I can only calculate AP using utils.compute_ap. I don't know how to calculate Average Recall!
> > 
> > 
> > you can calculate average recall from utils.compute_recall Also, compute_ap also returns iou, precision and recalls.
> 
> Could you please explain to me how to calculate mAR from the "utils.compute_recall" function, I understand that it returns the AR, but how should I calculate the mAR? Please help me!!

Hi @WillianaLeite ! Have you found a way to calculate mAR? I have the same issue here :)

**Comment by MALLI7622** [comment]
After fine-tuning Faster-RCNN on my custom data, I was getting this error
Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.000
Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=100 ] = 0.000
Average Precision (AP) @[ IoU=0.75 | area= all | maxDets=100 ] = 0.000
Average Precision (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
Average Precision (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
Average Precision (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 1 ] = 0.000
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 10 ] = 0.000
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.000
Average Recall (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
Average Recall (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
Average Recall (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000

@Suvi-dha @ashnair1  Can anybody help to resolve this issue...?

**Comment by keremtatlici** [comment]
@MALLI7622  Can you show us a sample of your dataset, i mean 1 sample's mask and original image

**Comment by Olamilekan002** [comment]
It seems the model is not learning at all

**Comment by MGardien** [comment]
> After fine-tuning Faster-RCNN on my custom data, I was getting this error Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.000 Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=100 ] = 0.000 Average Precision (AP) @[ IoU=0.75 | area= all | maxDets=100 ] = 0.000 Average Precision (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000 Average Precision (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000 Average Precision (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000 Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 1 ] = 0.000 Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 10 ] = 0.000 Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.000 Average Recall (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000 Average Recall (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000 Average Recall (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
> 
> @Suvi-dha @ashnair1 Can anybody help to resolve this issue...?

I had a similar issue. You should check the mask format. I could contain greycolors, where it should contain classes.

**Comment by babanthierry94** [comment]
> After fine-tuning Faster-RCNN on my custom data, I was getting this error Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.000 Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=100 ] = 0.000 Average Precision (AP) @[ IoU=0.75 | area= all | maxDets=100 ] = 0.000 Average Precision (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000 Average Precision (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000 Average Precision (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000 Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 1 ] = 0.000 Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 10 ] = 0.000 Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.000 Average Recall (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000 Average Recall (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000 Average Recall (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
> 
> @Suvi-dha @ashnair1 Can anybody help to resolve this issue...?

I have the same issue. Please did someone solve it ?

**Comment by Ann-ad** [comment]
@babanthierry94 @MGardien What is your number of training steps?

**Comment by Ann-ad** [comment]
@babanthierry94 @MGardien , Try increasing the number of your training steps

**Comment by nimakasipour** [comment]
@babanthierry94  the metric works if you perform the object detection on all images in the coco dataset, i found out when i saved the ground truth bounding boxes from "instances_val2017.json"  and tried to evaluate them with with these metrics (except to get 1.0 for all metrics). if some images are missing, then the result is lower or not 1.0, i used the fifty-one metrics which is also proposed on the homepage of the coco dataset, with fiftyone it was possible to evaluate the results even if they are not performed on all images

**Comment by andycga** [comment]
Can anybody help to resolve this issue...? 
Accumulating evaluation results...
DONE (t=0.23s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.988
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.991
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.991
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.988
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.993
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.993
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.993
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.993

what does the -1 mean?
